### PR TITLE
Fusion in Decoder (FiD)

### PR DESCRIPTION
**Pull request by mihaela-bornea**
# PrimeQA FiD Pull Request

## Adds support for generative QA and FiD model

Closes #(300)

## Description

**Describe** the changes proposed by this PR below to give the reviewer context below ↓↓↓

Components:
- Preprocessor: ELI5FiDPreprocessor
- FiD model: FiDModelForDownstreamTasks
- Data collator: FiDDataCollator
- Seq2Seq Trainer: MRCSeq2SeqTrainer
- Postprocessor: ELI5FiDPreprocessor
- Metric: Rouge

Most changes are in the run_mrc.py

To run train and evaluation:

```
python primeqa/primeqa/mrc/run_mrc.py \
--model_name_or_path facebook/bart-large \
--train_file /dccstor/mabornea1/lfqa_kb_data/kilt_eli5_dpr/eli5-train-kilt-dpr.json \
--eval_file /dccstor/mabornea1/lfqa_kb_data/kilt_eli5_dpr/eli5-dev-kilt-dpr.json \
--preprocessor primeqa.mrc.processors.preprocessors.eli5_fid.ELI5FiDPreprocessor \
--postprocessor primeqa.mrc.processors.postprocessors.eli5_fid.ELI5FiDPostProcessor \
--task_head primeqa.mrc.models.heads.generative.FID_HEAD \
--task_model primeqa.mrc.models.fid_task_model.FiDModelForDownstreamTasks \
--task_data_collator primeqa.mrc.data_models.data_collator.FiDDataCollator \
--task_trainer primeqa.mrc.trainers.seq2seq_mrc.MRCSeq2SeqTrainer \
--eval_metrics "ROUGE" \
--max_seq_length 256 \
--generation_num_beams 1 \
--generation_max_length 256 \
--max_answer_length 256 \
--predict_with_generate \
--max_contexts 3 \
--output_dir /dccstor/mabornea2/oneqa_os/primeqa_fid/exp_fid_fp16 \
--do_train --learning_rate 3e-5 --num_train_epochs 3 --fp16 \
--do_eval --per_device_train_batch_size 32 --per_device_eval_batch_size 32 \
--gradient_accumulation_steps 1 \
--overwrite_output_dir \
--overwrite_cache \
--evaluation_strategy epoch \
--save_strategy epoch \
--max_train_samples 100000
```

### Request Review

Be sure to **request a review** from one or more reviewers (unless the PR is to an unprotected branch).

## Versioning

When opening a PR to make changes to PrimeQA (i.e. `primeqa/`) master, be sure to increment the version following
[semantic versioning](https://semver.org/).  The VERSION is stored [here](https://github.com/primeqa/primeqa/blob/main/VERSION) 
and is incremented using `bump2version {patch,minor,major}` as described in the development guide documentation (https://github.com/primeqa/primeqa/blob/main/docs/development.md).

 - [ ] Have you updated the VERSION?
 - [ ] Or does this PR not change the `primeqa` package or was not into master?

After pulling in changes from master to an existing PR, ensure the VERSION is updated appropriately.
This may require bumping the version again if it has been previously bumped.

If you're not quite ready yet to post a PR for review, feel free to open a [draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

### After Merging
If merging into master and VERSION was updated, after this PR is merged:
- [ ] [Create a release](https://docs.github.com/en/github/administering-a-repository/releasing-projects-on-github/managing-releases-in-a-repository#creating-a-release) from the master with version equal to [VERSION](https://github.com/primeqa/primeqa/blob/main/VERSION)
- [ ] Not merging to master or VERSION not updated

## Checklist

Review the following and mark as completed:
- [x] [Tag an issue or issues](#what-does-this-pr-do) this PR addresses.
- [x] Added [description](#description) of changes proposed.
- [ ] Review requested as [appropriate](#request-review).
- [ ] Version bumped as [appropriate](#Versioning).
- [ ] New classes, methods, and functions documented.
- [ ] Documentation for modified code is updated.
- [ ] Built documentation to confirm it renders as expected (see https://github.com/primeqa/primeqa/blob/main/docs/development.md).
- [x] Code cleaned up and commented out code removed.
- [ ] Tests added to ensure all functionalities tested at >= 60% unit test coverage (see https://github.com/primeqa/primeqa/blob/main/docs/development.md).
- [ ] Code cleaned up and commented out code removed.
- [ ] Release created [as needed](#after-merging) after merging.

### CODE REVIEW

File: examples/lfqa/README.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
+
+# Long Form Question Answering (LFQA)
+
+LFQA is a form of generative question answering. Given a question,  the retriever component retrieves the supporting documents and a reader component generates the answer conditioned on the supporting documents.  The system generates complex multi-sentence answers. 
+
+## KILT-ELI5
+
+The following shows how to build information retrieval and reader components to generate answers for the KIlT-ELI5 dataset.
```
Comment (seirasto) [inline examples/lfqa/README.md:8]:
```suggestion
The following shows how to build information retrieval and reader components to generate answers for the KILT-ELI5 dataset.
```

File: examples/lfqa/README.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+```
+python examples/lfqa/kilt_passage_corpus.py \
+    --kilt_corpus $KILT_ELI5/data/kilt_knowledgesource.json \
+    --output_dir $KILT_ELI5/passages/ 
+    --passage_ids $KILT_ELI5/kilt_passage_ids.txt
+```
+
+### 3. Create a Dense Index with ColBERT
+
+At this point it is assumed that a ColBERT checkpoint already exists at `$COLBERT_CHECKPOINT`.
```
Comment (seirasto) [inline examples/lfqa/README.md:37]: Add a link here to IR readme in case this needs to be done?

File: examples/lfqa/README.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+python examples/lfqa/create_ir_queries_from_dataset.py \
+    --train_file $KILT_ELI5/data/eli5-train-kilt.jsonl \
+    --eval_file $KILT_ELI5/data/eli5-dev-kilt.jsonl \
+    --queries_per_file 50000 \
+    --output_dir $KILT_ELI5/queries \
+```
+
+### 4. Run the Search for the ELI5 Dataset
+
+The search is designed to run in parallel for chuncks of max 50000 queries.
```
Comment (seirasto) [inline examples/lfqa/README.md:68]:
```suggestion
The search is designed to run in parallel for chunks of max 50000 queries. 
```

File: examples/lfqa/README.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+```
+python examples/lfqa/add_passages_to_dataset.py \
+    --train_file $KILT_ELI5/data/eli5-train-kilt.jsonl \
+    --eval_file $KILT_ELI5/data/eli5-dev-kilt.jsonl \
+    --output_dir $KILT_ELI5/kilt-eli5-colbert-passages \
+    --search_result_location $KILT_ELI5/search_results \
+    --collection $KILT_ELI5/passages/kilt_knowledgesource_0.tsv \
+```
+
+### 6. Run the Reader Component on the Eli5 Dataset with Supporting Passages
```
Comment (seirasto) [inline examples/lfqa/README.md:110]:
```suggestion
### 6. Run the Reader Component on the ELI5 Dataset with Supporting Passages
```

File: examples/lfqa/README.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+The command below trains and evaluates an FiD generative reader model based on `facebook/bart-large`.<br>
+The number of supporting passages is 3. 
+The max length of the inout text is 256.
```
Comment (seirasto) [inline examples/lfqa/README.md:114]:
```suggestion
The max length of the input text is 256. 
```

File: examples/lfqa/README.md [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    --search_result_location $KILT_ELI5/search_results \
+    --collection $KILT_ELI5/passages/kilt_knowledgesource_0.tsv \
+```
+
+### 6. Run the Reader Component on the Eli5 Dataset with Supporting Passages
+
+The command below trains and evaluates an FiD generative reader model based on `facebook/bart-large`.<br>
+The number of supporting passages is 3. 
+The max length of the inout text is 256. 
+The max length of the generated anser is 256. <br>
```
Comment (seirasto) [inline examples/lfqa/README.md:115]:
```suggestion
The max length of the generated answer is 256. <br>
```

File: .github/workflows/primeqa-ci.yml [hunk 7]
```diff
@@ -42,9 +42,9 @@ jobs:
           echo "Run IR tests"
           pytest -s --cov primeqa/ir --cov-config .coveragerc --cov-fail-under=50 tests/primeqa/ir
           echo "Run MRC tests excluding trainer"
-          pytest -s --cov primeqa/mrc/processors --cov-config .coveragerc --cov-fail-under=50 --ignore tests/primeqa/mrc/trainers/test_mrc.py  tests/primeqa/mrc
```
Comment (seirasto) [inline .github/workflows/primeqa-ci.yml:45]: Why did this need to be changed?

File: examples/lfqa/add_passages_to_dataset.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+        help='the QA dev file',
+        type=str)
+    parser.add_argument('--test_file',  
+        help='the QA test file',
+        type=str)
+    parser.add_argument('--search_result_location',  
+        help='the output directory with the search result for the train dev and test queries',
+        type=str) 
+    parser.add_argument('--collection',  
+        help='The corpus file with all passages',
```
Comment (seirasto) [inline examples/lfqa/add_passages_to_dataset.py:22]:
```suggestion
        help='The corpus file with all the passages',
```

File: examples/lfqa/add_passages_to_dataset.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
+import argparse
```
Comment (seirasto) [inline examples/lfqa/add_passages_to_dataset.py:1]: This file could use some comments. Perhaps a description on the top describing what it does (and why)

File: notebooks/lfqa/retriever_reader_for_lfqa_predict_mode.ipynb [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    "colbert_retriever = ColBERTRetriever(index_root = index_root, index_name = index_name, collection = collection, max_num_documents = 3)\n",
+    "colbert_retriever.load()\n",
+    "\n",
+    "fid_reader = GenerativeFiDReader()\n",
+    "fid_reader.load()\n",
+    "\n",
+    "lfqa_pipeline = QAPipeline(colbert_retriever, fid_reader)\n",
+    "\n",
+    "questions = [\"What causes the trail behind jets at high altitude?\"]\n",
+    "answers = lfqa_pipeline.run(questions)"
```
Comment (seirasto) [inline notebooks/lfqa/retriever_reader_for_lfqa_predict_mode.ipynb:35]: Can you show some sample output in this file?

File: primeqa/ir/dense/colbert_top/colbert/evaluation/loaders.py [hunk 11]
```diff
@@ -19,8 +19,10 @@ def load_queries(queries_path):
 
     with open(queries_path) as f:
         for line in f:
+            if len(line.strip().split('\t')) < 2:
```
Comment (seirasto) [inline primeqa/ir/dense/colbert_top/colbert/evaluation/loaders.py:22]: Is this debugging that can be removed?

File: primeqa/ir/dense/colbert_top/colbert/evaluation/loaders.py [hunk 12]
```diff
@@ -19,8 +19,10 @@ def load_queries(queries_path):
...
+                print(line)
             qid, query, *_ = line.strip().split('\t')
-            qid = int(qid)
+            # qid = int(qid)
```
Comment (seirasto) [inline primeqa/ir/dense/colbert_top/colbert/evaluation/loaders.py:25]: remove?

File: primeqa/mrc/data_models/data_collator.py [hunk 13]
```diff
@@ -0,0 +1,69 @@
+from typing import Dict, List, Optional, Union, Any, Tuple
+from dataclasses import dataclass
+from transformers.file_utils import PaddingStrategy
+from transformers.tokenization_utils_base import PreTrainedTokenizerBase
+import torch
+
+# Modified from DataCollatorForSeq2Seq
+# convert feature list to tensor
+@dataclass
+class FiDDataCollator:
```
Comment (seirasto) [inline primeqa/mrc/data_models/data_collator.py:10]: please add docstring

File: primeqa/pipelines/README.md [hunk 14]
```diff
@@ -98,4 +99,88 @@ as the seat of government for the new nation, it is Australia's
 largest inland city"""]
 answers = reader.apply(question,context)  
 print(json.dumps(answers, indent=4))  
+```
+
+
+### Generative FiD Reader
+
+A Generative Reader takes a question and uses a set of supporting passages to qenerate an answer. Compared to the Extractive Reader, where the answers are usually short spans extracted from the input passages, the Generative Reader generates complex, multi-sentence answers.
```
Comment (seirasto) [inline primeqa/pipelines/README.md:107]:
```suggestion
A Generative Reader takes a question and uses a set of supporting passages to generate an answer. In contrast to the Extractive Reader, where the answers are usually short spans extracted from the input passages, the Generative Reader generates complex, multi-sentence answers.
```

File: primeqa/pipelines/README.md [hunk 15]
```diff
@@ -98,4 +99,88 @@ as the seat of government for the new nation, it is Australia's
...
+print(json.dumps(answers, indent=4)) 
+```
+
+## QA Pipeline
+
+The QA pipeline is used for Open Retrieval Question Answering. Open retrieval systems query large document stores for relevant passages. Long Form Question Answering (LFQA) is a generative task where the retrieved passages are used to generate a complex multi-sentence answer.
+
+In this example we show a QA Pipeline using a ColBERT retriever  and a Fusion in Decoder (FID) generator.
+
+Instructions to create a ColBERT index and an FiD model for KILT-ELI5 can be found [here]()
```
Comment (seirasto) [inline primeqa/pipelines/README.md:160]: missing link

File: primeqa/pipelines/README.md [hunk 16]
```diff
@@ -98,4 +99,88 @@ as the seat of government for the new nation, it is Australia's
...
+
+- Step 1:  Initialize the retriever.
+
+```python
+retriever = ColBERTRetriever(index_root = index_root, index_name = index_name, collection = collection, max_num_documents = 3)
+```
+
+- Step 2:  Initialize the reader model. You can choose any the generative QA model we currently have ...
```
Comment (seirasto) [inline primeqa/pipelines/README.md:168]:
```suggestion
- Step 2:  Initialize the reader model. You can choose any generative QA model we currently have ...
```

File: tests/primeqa/mrc/metrics/rouge/test_rouge.py [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+class TestROUGE(UnitTest):
+    
+    @pytest.fixture(scope='session')
+    def metric(self):
+        return ROUGE()
+    
+    @pytest.fixture(scope='session')
+    def references_and_predictions(self):
+        references = [
+            dict(id='1', answers=["hello there"]),
```
Comment (seirasto) [inline tests/primeqa/mrc/metrics/rouge/test_rouge.py:16]: add a test when not equal?

**Review by seirasto — COMMENTED** [review]
Looks good for the most part. I've left minor comments in many files. I think docstring could be used in more of the classes and comments in the examples folder files (the scripts for setup of FiD). I will also suggest that we have a layer of folders for extractive vs generative.

File: .github/workflows/primeqa-ci.yml (hunk 7 above)
Comment (mihaela-bornea) [reply .github/workflows/primeqa-ci.yml:45]: Because the CI was not running the new unit test but was reporting the coverage was low.

File: examples/lfqa/README.md (hunk 3 above)
Comment (mihaela-bornea) [reply examples/lfqa/README.md]: Done

File: tests/primeqa/mrc/metrics/rouge/test_rouge.py (hunk 17 above)
Comment (mihaela-bornea) [reply tests/primeqa/mrc/metrics/rouge/test_rouge.py:16]: We decided this test is sufficient for now.

**Review by bhavani105 — COMMENTED** [review]
Should  primeqa/mrc/README.md be updated with the additional arguments and the available options.

File: primeqa/mrc/data_models/data_collator.py [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+from transformers.file_utils import PaddingStrategy
+from transformers.tokenization_utils_base import PreTrainedTokenizerBase
+import torch
+
+
+@dataclass
+class FiDDataCollator:
+    r"""
+    Modified from DataCollatorForSeq2Seq
+    Do not pad the features for FID
```
Comment (seirasto) [inline primeqa/mrc/data_models/data_collator.py:12]:
```suggestion
    Do not pad the features for FiD 
```

File: primeqa/mrc/models/fid_task_model.py [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+from transformers.modeling_outputs import BaseModelOutput
+
+from primeqa.mrc.models.heads.abstract import AbstractTaskHead
+
+class EncoderWrapper(torch.nn.Module):
+    """
+    EncoderWrapper for the FiD model
+    
+    B - Batch size
+    N the number of passages per example
```
Comment (seirasto) [inline primeqa/mrc/models/fid_task_model.py:15]:
```suggestion
    N - the number of passages per example
```

File: primeqa/mrc/models/fid_task_model.py [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+    L the max seq length
```
Comment (seirasto) [inline primeqa/mrc/models/fid_task_model.py:16]:
```suggestion
    L - the max seq length
```

File: primeqa/mrc/processors/postprocessors/eli5_fid.py [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+
+        # Build a map example to its corresponding features.
+        example_id_to_index = {k: i for i, k in enumerate(examples["id"])} 
+
+        feature_per_example = {example_id_to_index[feature["example_id"]]: i for i, feature in enumerate(features)}
+        predictions = {}
+
+        # Let's loop over all the examples!
+        for example_index, example in enumerate(examples):
+            # This is the index of the feature associated to the current example.
```
Comment (seirasto) [inline primeqa/mrc/processors/postprocessors/eli5_fid.py:39]:
```suggestion
            # This is the index of the feature associated with the current example.
```

File: primeqa/pipelines/README.md [hunk 22]
```diff
@@ -98,4 +99,89 @@ as the seat of government for the new nation, it is Australia's
 largest inland city"""]
 answers = reader.apply(question,context)  
 print(json.dumps(answers, indent=4))  
+```
+
+
+### Generative FiD Reader
+
+A Generative Reader takes a question and uses a set of supporting passages to generate an answer. In contrast to the Extractive Reader, where the answers are usually short spans extracted from the input passages, the Generative Reader generates complex, multi-sentence answers.
```
Comment (seirasto) [inline primeqa/pipelines/README.md:107]:
```suggestion
A Generative Reader takes a question and a set of supporting passages (FiD usually uses 3 supporting passage) to generate an answer. In contrast to the Extractive Reader, where the answers are usually short spans extracted from the input passages, the Generative Reader generates complex, multi-sentence answers.
```

File: primeqa/pipelines/components/reader/generative.py [hunk 23]
```diff
@@ -24,3 +32,118 @@ def load(self, *args, **kwargs):
...
+    )
+    max_seq_len: int = field(
+        default=256,
+        metadata={
+            "name": "Maximum sequence length",
+            "description": "Maximum length of question and context inputs to the model (in word pieces/bpes)",
+            "range": [32, 512, 8],
+        },
+    )
+    max_answer_length: int = field(
```
Comment (seirasto) [inline primeqa/pipelines/components/reader/generative.py:51]: What is the difference between max_answer_length and generation_max_length?

File: primeqa/pipelines/components/reader/generative.py [hunk 24]
```diff
@@ -24,3 +32,118 @@ def load(self, *args, **kwargs):
...
+            max_seq_len=self.max_seq_len,
+            tokenizer=tokenizer,
+            max_contexts=3,# self.num_contexts,
+            max_answer_len=self.generation_max_length
+        )
+
+        data_collator = FiDDataCollator(tokenizer)
+        
+         # Initialize post processor
+        postprocessor = ELI5FiDPostProcessor(
```
Comment (seirasto) [inline primeqa/pipelines/components/reader/generative.py:100]: I guess this is ok for now, but what if you wanted to use a new preprocessor? There is no ability to have a different one right now.

**Review by seirasto — APPROVED** [review]
LGTM. I've left a few more minor comments (mostly typo fixes). Please fix as you see appropriate.

File: primeqa/pipelines/components/reader/generative.py (hunk 23 above)
Comment (mihaela-bornea) [reply primeqa/pipelines/components/reader/generative.py:51]: They are redundant but the max_answer_length was used in the postprocessor while the trainer needs this generation_max_length which is a trainer argument from HF.

File: examples/lfqa/README.md [hunk 25]
```diff
@@ -0,0 +1,156 @@
...
+ Store the data into the `$KILT_ELI5/data` directory.
+
+The KILT knowledge source (the corpus) can be downloaded here: [kilt_knowledgesource.json](http://dl.fbaipublicfiles.com/KILT/kilt_knowledgesource.json) (34.76GiB).<br>
+It is based on the [2019/08/01 Wikipedia dump](http://dl.fbaipublicfiles.com/BLINK/enwiki-pages-articles.xml.bz2).<br>
+
+The train set of KILT-ELI5 can be downloaded from [eli5-train-kilt.jsonl](http://dl.fbaipublicfiles.com/KILT/eli5-train-kilt.jsonl) (272,634 lines, 548MiB) 
+
+The dev set of KILT-ELI5 can be downloaded from [eli5-dev-kilt.jsonl](http://dl.fbaipublicfiles.com/KILT/eli5-dev-kilt.jsonl) (1,507 lines, 14.1MiB) 
+
+### 2. Preprocess the KILT Knowledge Surce
```
Comment (bhavani105) [inline examples/lfqa/README.md:24]: Preprocess the KILT Knowledge Source
```suggestion
### 2. Preprocess the KILT Knowledge Source
```

File: examples/lfqa/README.md [hunk 26]
```diff
@@ -0,0 +1,156 @@
...
+    --do_eval --per_device_train_batch_size 32 --per_device_eval_batch_size 32 \
+    --gradient_accumulation_steps 1 \
+    --fp16 \
+    --evaluation_strategy epoch \
+    --save_strategy epoch \
+    --load_best_model_at_end \
+    --metric_for_best_model rougeL \
+    --overwrite_output_dir \
+    --overwrite_cache \
+```
```
Comment (bhavani105) [inline examples/lfqa/README.md:156]: A shell script that runs all these steps in sequence would make is easier.

**Review by bhavani105 — APPROVED** [review]
LGTM
Left some comments
- simplify the ELI5 example steps by combining all the steps from downloading data to building the index into one shell script
- minor edits

File: primeqa/pipelines/components/reader/generative.py (hunk 23 above)
Comment (seirasto) [reply primeqa/pipelines/components/reader/generative.py:51]: Can you just set one using the other in the init function?

File: primeqa/mrc/models/fid_task_model.py [hunk 27]
```diff
@@ -0,0 +1,151 @@
...
+                input_ids=None,
+                attention_mask=None,
+                labels=None,
+                return_dict=False,
+                **kwargs):
+    
+        if input_ids != None:
+            if input_ids.dim() == 3:
+                encoder = self.get_encoder()
+                encoder.n_passages = input_ids.size(1)
```
Comment (rongzhangibm) [inline primeqa/mrc/models/fid_task_model.py:82]: How to set encoder.n_passages if input_ids.dim() != 3 ?

File: primeqa/mrc/models/fid_task_model.py [hunk 28]
```diff
@@ -0,0 +1,151 @@
...
+            attention_mask=attention_mask,
+            labels=labels,
+            return_dict=return_dict,
+            **kwargs
+        ) # tuple: (loss, lm_logits,) + BartModel outputs
+        return outputs
+        
+    def generate(self, input_ids, **gen_kwargs):
+        encoder = self.get_encoder()
+        encoder.n_passages = input_ids.size(1)
```
Comment (rongzhangibm) [inline primeqa/mrc/models/fid_task_model.py:97]: Do we need to check input_ids.dim() here?

File: primeqa/mrc/processors/postprocessors/eli5_fid.py [hunk 29]
```diff
@@ -0,0 +1,59 @@
...
+         # Decode the predicted tokens.
+        preds = predictions.predictions
+        if isinstance(preds, tuple):
+            preds = preds[0]
+        decoded_preds = self.tokenizer.batch_decode(preds, skip_special_tokens=True)
+
+        # Build a map example to its corresponding features.
+        example_id_to_index = {k: i for i, k in enumerate(examples["id"])} 
+
+        feature_per_example = {example_id_to_index[feature["example_id"]]: i for i, feature in enumerate(features)}
```
Comment (rongzhangibm) [inline primeqa/mrc/processors/postprocessors/eli5_fid.py:34]: Does this work for the case that one example is mapped to multi features? or this never happens?

File: primeqa/mrc/processors/postprocessors/eli5_fid.py [hunk 30]
```diff
@@ -0,0 +1,59 @@
...
+        predictions = {}
+
+        # Let's loop over all the examples!
+        for example_index, example in enumerate(examples):
+            # This is the index of the feature associated to the current example.
+            feature_index = feature_per_example[example_index]
+            predictions[example["id"]] = decoded_preds[feature_index]
```
Comment (rongzhangibm) [inline primeqa/mrc/processors/postprocessors/eli5_fid.py:41]: Seems one example can only generate one feature?

File: primeqa/mrc/processors/preprocessors/eli5_fid.py [hunk 31]
```diff
@@ -0,0 +1,147 @@
...
+        indices = []
+        questions = examples[self._question_column]
+        contexts = examples[self._context_column]
+        n_doc = self._max_contexts
+
+        def top_passages(ctx):
+            assert n_doc <= len(ctx) 
+            return [ctx[i]["text"] for i in range(n_doc)]
+        def append_question(passages, question):
+            return [f"question: {question} passage: {t}" for t in passages]
```
Comment (rongzhangibm) [inline primeqa/mrc/processors/preprocessors/eli5_fid.py:105]: Just curious why use the format? How about {question} <sep> {passage} ?

File: primeqa/mrc/trainers/seq2seq_mrc.py [hunk 32]
```diff
@@ -0,0 +1,134 @@
...
+        self.post_process_function = post_process_function
+
+    def evaluate(
+        self,
+        eval_dataset: Optional[Dataset] = None,
+        eval_examples=None,
+        ignore_keys: Optional[List[str]] = None,
+        metric_key_prefix: str = "eval",
+        max_length: Optional[int] = None,
+        num_beams: Optional[int] = None,
```
Comment (rongzhangibm) [inline primeqa/mrc/trainers/seq2seq_mrc.py:54]: Can we also add 'do_sample' to the input argument to support sampling instead of greedy search in generation?

File: primeqa/pipelines/components/reader/generative.py [hunk 33]
```diff
@@ -24,3 +32,118 @@ def load(self, *args, **kwargs):
...
+    generation_max_length: int = field(
+        default=256, metadata={"name": "Maximum answer length", "range": [2, 2000, 2]}
+    )
+    generation_num_beams: int = field(
+        default=1, metadata={"name": "The number of beams for generation", "range": [1, 5, 1]}
+    )
+    num_contexts: int = field(
+        default=3, metadata={"name": "The number of passages in the input", "range": [1, 10, 1]}
+    )
+
```
Comment (rongzhangibm) [inline primeqa/pipelines/components/reader/generative.py:63]: 'do_sample' / 'do_greedy_search' ?

**Review by rongzhangibm — APPROVED** [review]
LGTM. Left some comments.

**Comment by mihaela-bornea** [comment]
@bhavani105 :  I will postpone adding documentation on how to used custom reader, trainer until a decision is made about the direction of V2. Also, more discussion should go about where we should document all the related arguments. Right now we talk about preprocessor, but we did not described the custom postprocessor, metric, etc.
I will also leave the scripts in the LFQA separate for now. I think it might be beneficial for someone to understand exactly where the data is coming from. 

@rongzhangibm : I think it is a good idea to add support for other generation strategies but I prefer to do a separate PR on this.
